core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesCommitUtil.java

```diff
@@ -0,0 +1,208 @@
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Functionality used by RewriteDataFile Actions from different platforms to handle commits.
+ */
+public class RewriteDataFilesCommitUtil {
+  private static final Logger LOG = LoggerFactory.getLogger(RewriteDataFilesCommitUtil.class);
+
+  private final Table table;
+
+  public RewriteDataFilesCommitUtil(Table table) {
+    this.table = table;
+  }
+
+  /**
+   * Perform a commit operation on the table adding and removing files as
+   * required for this set of file groups
+   * @param fileGroups fileSets to commit
+   */
+  public void commitFileGroups(Set<RewriteFileGroup> fileGroups) {
+    Set<DataFile> rewrittenDataFiles = fileGroups.stream()
+        .flatMap(fileGroup -> fileGroup.rewrittenFiles().stream()).collect(Collectors.toSet());
+
+    Set<DataFile> newDataFiles = fileGroups.stream()
+        .flatMap(fileGroup -> fileGroup.addedFiles().stream()).collect(Collectors.toSet());
+
+    table.newRewrite()
+        .rewriteFiles(rewrittenDataFiles, newDataFiles)
+        .commit();
+  }
+
+  /**
+   * Clean up a specified file set by removing any files created for that operation, should
+   * not throw any exceptions
+   * @param fileGroup fileSet to clean
+   */
+  public void abortFileGroup(RewriteFileGroup fileGroup) {
+    if (fileGroup.addedFiles() == null) {
+      // Nothing to clean
+      return;
+    }
+    Tasks.foreach(fileGroup.addedFiles())
+        .noRetry()
+        .suppressFailureWhenFinished()
+        .onFailure((dataFile, exc) -> LOG.warn("Failed to delete: {}", dataFile.path(), exc))
+        .run(dataFile -> table.io().deleteFile(dataFile.path().toString()));
+  }
+
+  public void commitOrClean(Set<RewriteFileGroup> rewriteGroups) {
+    try {
+      commitFileGroups(rewriteGroups);
+    } catch (CommitStateUnknownException e) {
+      LOG.error("Commit state unknown for {}, cannot clean up files because they may have been committed successfully.",
+          rewriteGroups, e);
+      throw e;
+    } catch (Exception e) {
+      LOG.error("Cannot commit groups {}, attempting to clean up written files", rewriteGroups, e);
+      Tasks.foreach(rewriteGroups)
+          .suppressFailureWhenFinished()
+          .run(group -> abortFileGroup(group));
+      throw e;
+    }
+  }
+
+  /**
+   * An async service which allows for committing multiple file groups as their rewrites complete. The service also
+   * allows for partial-progress since commits can fail. Once the service has been closed no new file groups should not
+   * be offered.
+   * @param rewritesPerCommit number of file groups to include in a commit
+   * @return the service for handling commits
+   */
+  public CommitService service(int rewritesPerCommit) {
+    return new CommitService(rewritesPerCommit);
+  }
+
+  public class CommitService implements Closeable {
```

Comment:
Could this be a static inner class instead?

Reply:
Actually I ran into some worries here, I'm a little nervous about the pieces of this class that have state, like the table and executor service (and queue). I think we are safer if it's not static.